src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs

```diff
@@ -1090,36 +1259,168 @@ private void CompileSourceToAssembly(string source)
+                    }
+                }
+
+                if (IsError)
+                {
+                    ErrorRecord errorRecord = new ErrorRecord(
+                        new InvalidOperationException(AddTypeStrings.CompilerErrors),
+                        "COMPILER_ERRORS",
+                        ErrorCategory.InvalidData,
+                        null);
+                    ThrowTerminatingError(errorRecord);
+                }
             }
         }
 
-        private AddTypeCompilerError[] GetErrors(ImmutableArray<Diagnostic> diagnostics)
+        private string BuildErrorMessage(Diagnostic diagnisticRecord)
         {
-            return diagnostics.Where(d => d.Severity == DiagnosticSeverity.Error || d.Severity == DiagnosticSeverity.Warning)
-                .Select(d => new AddTypeCompilerError
+            var location = diagnisticRecord.Location;
+
+            if (location.SourceTree == null)
+            {
+                // For some error types (linker?) we don't have related source code.
+                return diagnisticRecord.ToString();
+            }
+            else
+            {
+                var text = location.SourceTree.GetText();
+                var textLines = text.Lines;
+
+                var lineSpan = location.GetLineSpan(); // FileLinePositionSpan type.
+                var errorLineNumber = lineSpan.StartLinePosition.Line;
+                var errorPosition = lineSpan.StartLinePosition.Character;
+
+                // This is typical Roslyn diagnostic message which contains
+                // a message number, a source context and an error position.
+                var diagnisticMessage = diagnisticRecord.ToString();
+                var errorLineString = textLines[errorLineNumber].ToString();
+
+                StringBuilder sb = new StringBuilder(diagnisticMessage.Length + errorLineString.Length * 2 + 4);
+
+                sb.AppendLine(diagnisticMessage);
+                sb.AppendLine(errorLineString);
+
+                for (var i = 0 ; i < errorLineString.Length ; i++)
                 {
-                    ErrorText = d.GetMessage(),
-                    FileName = null,
-                    Line = d.Location.GetMappedLineSpan().StartLinePosition.Line + 1, // Convert 0-based to 1-based
-                    IsWarning = !d.IsWarningAsError && d.Severity == DiagnosticSeverity.Warning,
-                    Column = d.Location.GetMappedLineSpan().StartLinePosition.Character + 1, // Convert 0-based to 1-based
-                    ErrorNumber = d.Id
-                }).ToArray();
+                    if (errorLineString[i] != ' ' && errorLineString[i] != '\t')
+                    {
+                        // We copy white chars from the source string.
+                        sb.Append(errorLineString, 0, i);
+                        // then pad up to the error position.
+                        sb.Append(' ', Math.Max(0, lineSpan.StartLinePosition.Character - i));
```

Comment:
> Math.Max(0, lineSpan.StartLinePosition.Character - i)

Is the `Math.Max` call really needed? It seems to me that `errorPosition` would be `-ge i`.

Reply:
When I researched how to make the error report I was looking for code examples. Now I do not remember where I saw this pattern.
I think this is because the lineSpan and the error position are completely different code paths - so we should protect.